File: scripts/fix-puppeteer-recursion.js
const fs = require('fs')
const path = 'repo/server/lib/cf-harvester/puppeteer.js'
let text = fs.readFileSync(path, 'utf8')

const startMarker = '  async function _extractStreamImpl(watchUrl, options = {}) {'
const start = text.indexOf(startMarker)
if (start === -1) { console.log('start not found')
process.exit(1) }

const endMatch = text.search(/\n  }\n\n  async function exportCookiesImpl/)
if (endMatch === -1) { console.log('end not found')
process.exit(1) }
const end = endMatch + 4

let func = text.slice(start, end)

const split1 = '    // ── Gogoanime retry loop with proxy rotation ─────────────────────'
const idx1 = func.indexOf(split1)
if (idx1 === -1) { console.log('split1 not found')
process.exit(1) }
const head = func.slice(0, idx1)
const midTail = func.slice(idx1)

const split2 = '    console.log(`[cf-harvester] DOM extraction: ${watchUrl.slice(0, 100)}`)'
const idx2 = midTail.indexOf(split2)
if (idx2 === -1) { console.log('split2 not found')
process.exit(1) }
const retryPart = midTail.slice(0, idx2)
const logicPart = midTail.slice(idx2)

let wrapper = head.replace('async function _extractStreamImpl', 'async function _doExtractStream')
wrapper = wrapper.replace('return await _extractStreamImpl(watchUrl, options)', 'return await __doExtract(watchUrl, options)')
wrapper = wrapper + retryPart + '\n  }'

const helper = '  async function __doExtract(watchUrl, options = {}) {\n    const totalBudgetMs = options.maxDurationMs ?? 30_000\n    const remainingBudget = makeRemainingBudget(Date.now(), totalBudgetMs)\n    const isGogo = watchUrl.includes(\'gogoanime\')\n' + logicPart

const newFunc = wrapper + '\n\n' + helper
let newText = text.slice(0, start) + newFunc + text.slice(end)
newText = newText.replace(/_extractStreamImpl\(watchUrl, options\)/g, '_doExtractStream(watchUrl, options)')
fs.writeFileSync(path, newText, 'utf8')
console.log('fixed recursion')

File: scripts/fix-puppeteer-recursion.py
import re

path = 'repo/server/lib/cf-harvester/puppeteer.js'
with open(path, 'r', encoding='utf-8') as f:
    text = f.read()

start_marker = '  async function _extractStreamImpl(watchUrl, options = {}) {'
start = text.find(start_marker)
if start == -1:
    print('start not found')
    raise SystemExit(1)

end_match = re.search(r'\n  }\n\n  async function exportCookiesImpl', text)
if not end_match:
    print('end not found')
    raise SystemExit(1)
end = end_match.start() + len('\n  }')

func = text[start:end]

split1 = '    // ── Gogoanime retry loop with proxy rotation ─────────────────────'
idx1 = func.find(split1)
if idx1 == -1:
    print('split1 not found')
    raise SystemExit(1)
head = func[:idx1]
mid_tail = func[idx1:]

split2 = '    console.log(`[cf-harvester] DOM extraction: ${watchUrl.slice(0, 100)}`)'
idx2 = mid_tail.find(split2)
if idx2 == -1:
    print('split2 not found')
    raise SystemExit(1)
retry_part = mid_tail[:idx2]
logic_part = mid_tail[idx2:]

wrapper = head.replace('async function _extractStreamImpl', 'async function _doExtractStream')
wrapper = wrapper.replace('return await _extractStreamImpl(watchUrl, options)', 'return await __doExtract(watchUrl, options)')
wrapper = wrapper + retry_part + '\n  }'

helper = '  async function __doExtract(watchUrl, options = {}) {\n    const totalBudgetMs = options.maxDurationMs ?? 30_000\n    const remainingBudget = makeRemainingBudget(Date.now(), totalBudgetMs)\n    const isGogo = watchUrl.includes(\'gogoanime\')\n' + logic_part

new_func = wrapper + '\n\n' + helper
new_text = text[:start] + new_func + text[end:]
new_text = new_text.replace('_extractStreamImpl(watchUrl, options)', '_doExtractStream(watchUrl, options)')

with open(path, 'w', encoding='utf-8') as f:
    f.write(new_text)
print('fixed recursion')
